util/vec-map/src/lib.rs

```diff
@@ -0,0 +1,300 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Provides a map (key-value store) interface backed by `heapless::Vec`
+
+#![deny(missing_docs)]
+
+use displaydoc::Display;
+use heapless::Vec;
+
+/// An error which can occur when using VecMap
+#[derive(Clone, Debug, Display, Eq, PartialEq)]
+pub enum Error {
+    /// VecMap capacity exceeded
+    CapacityExceeded,
+}
+
+/// This is a mini version of VecMap that is no_std compatible and uses
+/// heapless::Vec instead alloc::Vec.
+///
+/// It may be better to patch upstream for no_std compatibility and use that,
+/// but that crate has other issues -- it relies on an
+/// experimental "contracts" crate that causes a dependency on rand crate.
+/// Porting to Heapless would be a breaking change to the API.
+///
+/// TBD what the best path is: https://github.com/p-avital/vec-map-rs/blob/master/src/lib.rs
+#[derive(Clone, Debug)]
+pub struct VecMap<K, V, const N: usize> {
+    keys: Vec<K, N>,
+    values: Vec<V, N>,
+}
+
+impl<K, V, const N: usize> Default for VecMap<K, V, N> {
+    fn default() -> Self {
+        Self {
+            keys: Default::default(),
+            values: Default::default(),
+        }
+    }
+}
+
+impl<K, V, const N: usize> VecMap<K, V, N> {
+    /// Check if the map is empty
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Get the number of items in the map
+    #[inline]
+    pub fn len(&self) -> usize {
+        debug_assert!(self.keys.len() == self.values.len());
+        self.keys.len()
+    }
+}
+
+impl<K: Eq + PartialEq, V, const N: usize> VecMap<K, V, N> {
+    /// Get the value associated to a key, if present
+    pub fn get(&self, key: &K) -> Option<&V> {
+        self.keys
+            .iter()
+            .position(|k| k == key)
+            .map(|idx| &self.values[idx])
+    }
+
+    /// Get a mutable reference to the value associated to a key, if present
+    pub fn get_mut(&mut self, key: &K) -> Option<&mut V> {
+        self.keys
+            .iter()
+            .position(|k| k == key)
+            .map(|idx| &mut self.values[idx])
+    }
+}
+
+impl<K: Clone + Eq + PartialEq, V, const N: usize> VecMap<K, V, N> {
+    /// Get a mutable reference to the value associated to a key, if present,
+    /// or else insert such a value produced by given callback,
+    /// and then return a mutable reference
+    ///
+    /// Returns an error if the heapless::Vec capacity was exceeded
+    pub fn get_mut_or_insert_with(
+        &mut self,
+        key: &K,
+        val_fn: impl FnOnce() -> V,
+    ) -> Result<&mut V, Error> {
+        if let Some(idx) = self.keys.iter().position(|k| k == key) {
+            Ok(&mut self.values[idx])
+        } else {
+            let idx = self.keys.len();
+            debug_assert_eq!(idx, self.values.len());
+            self.keys
+                .push(key.clone())
+                .map_err(|_| Error::CapacityExceeded)?;
+            self.values
```

Comment:
In some places we have debug_asserts that ensure the length of `keys` and `values` are the same, but in some places we don't (e.g. `get()`). If we were able to push into keys, and somehow failed to push into values, that might be worth of panicing since we are leaving the object in a funky state (and right now we will panic anyway if `.get()` is called on the key when something like this has happened).
Alternatively, maybe if this fails then popping the key is the right thing to do?

Reply:
+1 for popping the key if we cannot append the value

Reply:
Thanks